src/connectors/impls/gpubsub/gsub.rs

```diff
@@ -0,0 +1,331 @@
+use async_std::task::JoinHandle;
+use dashmap::DashMap;
+use googapis::google::pubsub::v1::subscriber_client::SubscriberClient;
+use googapis::google::pubsub::v1::{PubsubMessage, ReceivedMessage, StreamingPullRequest};
+use gouth::Token;
+use serde::Deserialize;
+use std::fmt::Debug;
+use std::sync::Arc;
+use std::time::Duration;
+use tonic::codegen::InterceptedService;
+use tonic::transport::{Certificate, Channel, ClientTlsConfig};
+use tonic::Status;
+use tremor_pipeline::ConfigImpl;
+
+#[derive(Deserialize, Clone)]
+struct Config {
+    pub connect_timeout: u64,
+    pub ack_deadline: u64,
+    pub subscription_id: String,
+    #[serde(default = "default_endpoint")]
+    pub endpoint: String,
+    #[serde(default = "default_skip_authentication")]
+    pub skip_authentication: bool,
+}
+impl ConfigImpl for Config {}
+
+fn default_endpoint() -> String {
+    "https://pubsub.googleapis.com".into()
+}
+
+fn default_skip_authentication() -> bool {
+    false
+}
+
+#[derive(Debug, Default)]
+pub(crate) struct Builder {}
+
+#[async_trait::async_trait]
+impl ConnectorBuilder for Builder {
+    fn connector_type(&self) -> ConnectorType {
+        "gsub".into()
+    }
+
+    async fn build(&self, alias: &str, raw_config: &ConnectorConfig) -> Result<Box<dyn Connector>> {
+        if let Some(raw) = &raw_config.config {
+            let config = Config::new(raw)?;
+            Ok(Box::new(GSub { config }))
+        } else {
+            Err(ErrorKind::MissingConfiguration(alias.to_string()).into())
+        }
+    }
+}
+
+struct GSub {
+    config: Config,
+}
+
+type PubSubClient = SubscriberClient<InterceptedService<Channel, AuthInterceptor>>;
+type AsyncTaskMessage = Result<(u64, Vec<u8>)>;
+
+struct GSubSource {
+    config: Config,
+    client: Option<PubSubClient>,
+    receiver: Option<Receiver<AsyncTaskMessage>>,
+    ack_sender: Option<Sender<u64>>,
+    ack_ids: Arc<DashMap<u64, String>>,
+    task_handle: Option<JoinHandle<()>>,
+}
+
+impl GSubSource {
+    pub fn new(config: Config) -> Self {
+        GSubSource {
+            config,
+            client: None,
+            receiver: None,
+            ack_ids: Arc::new(DashMap::new()),
+            task_handle: None,
+            ack_sender: None,
+        }
+    }
+}
+
+async fn consumer_task(
+    mut client: PubSubClient,
+    sender: Sender<AsyncTaskMessage>,
+    ack_ids: Arc<DashMap<u64, String>>,
+    subscription_id: String,
+    ack_deadline: Duration,
+    ack_receiver: Receiver<u64>,
+) {
+    let mut ack_counter = 0;
+
+    let ack_ids_clone = ack_ids.clone();
+    let request_stream = async_stream::stream! {
+        yield StreamingPullRequest {
+            subscription: subscription_id.clone(),
+            ack_ids: vec![],
+            modify_deadline_seconds: vec![],
+            modify_deadline_ack_ids: vec![],
+            stream_ack_deadline_seconds: i32::try_from(ack_deadline.as_secs()).unwrap_or(10),
+            client_id: "".to_string(),
```

Comment:
In kafka, we do assign our own client id: https://github.com/tremor-rs/tremor-runtime/blob/main/src/connectors/impls/kafka/consumer.rs#L92 should we do the same here?